<a href="https://colab.research.google.com/github/nmfairuz/ML-Coding/blob/main/Recommender_System_ContentBased_Filtering.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Recommender System ContentBased Filtering

### Collaborative Filtering

> "Recommend based on **similar users**."

Example:

* Wasi likes A and B.
* Another user likes A, B, and C.
* Recommend C.


### Content-Based Filtering

> "Recommend based on **similar items**."

Example:

You like:

* Action
* Sci-Fi
* Adventure

Movie:

* Action ✅
* Sci-Fi ✅
* Adventure ✅

Recommend it.

Notice we never looked at other users.


**Example Dataset :**

Let's pretend these are movies.

| Movie      | Action | Romance | Comedy |
| ---------- | ------ | ------- | ------ |
| Avengers   | 1      | 0       | 0      |
| Titanic    | 0      | 1       | 0      |
| Deadpool   | 1      | 0       | 1      |
| La La Land | 0      | 1       | 1      |

These are the **features**.

---

Now suppose **your preferences** are

```python
User = [0.9, 0.2, 0.4]

# Action
# Romance
# Comedy
```

Meaning

* Loves action
* Doesn't care much about romance
* Likes comedy a little


# TensorFlow/Keras version (Andrew later)

Sometimes you'll see two neural networks.

```text
Movie Features
        │
        ▼
Movie Network
        │
 Movie Embedding
        │
        │
        ▼
Dot Product
        ▲
        │
 User Embedding
        │
User Network
        ▲
User Features
```

Each network learns embeddings.

Recommendation score = Dot Product

Exactly like collaborative filtering, except here **both user and item have known features**.


```text
# Content-Based Filtering

Goal:
Recommend items similar to what the user already likes.

Idea:
Use item features instead of other users' preferences.

Example:
If a user likes Action and Sci-Fi movies,
recommend movies with similar features.

Workflow:

Item Features
        ↓
User Preferences
        ↓
Similarity Score (Dot Product)
        ↓
Recommend highest-scoring items

Advantages:
✔ Works for new items.
✔ Personalized recommendations.
✔ Doesn't depend on other users.

Disadvantages:
✘ Requires good item features.
✘ May recommend only similar content (less diversity).
```

## import

In [1]:
import tensorflow as tf
import numpy as np

## Movie features

In [2]:
movie_features = tf.constant([
    [1,0,0],   # Avengers
    [0,1,0],   # Titanic
    [1,0,1],   # Deadpool
    [0,1,1]    # La La Land
], dtype=tf.float32)

## User preferences

In [3]:
user_preferences = tf.constant(
    [0.9,0.2,0.4],
    dtype=tf.float32
)

## Similarity score

This is just a dot product.

In [4]:
scores = tf.linalg.matvec(
    movie_features,
    user_preferences
)

print(scores.numpy())

[0.9 0.2 1.3 0.6]


| Movie      | Score |
| ---------- | ----- |
| Avengers   | 0.9   |
| Titanic    | 0.2   |
| Deadpool   | 1.3   |
| La La Land | 0.6   |


Recommendation:

Deadpool ⭐⭐⭐⭐⭐

Avengers ⭐⭐⭐⭐

La La Land ⭐⭐⭐

Titanic ⭐